rust/src/guess.rs

```diff
@@ -105,64 +161,95 @@ impl Guess {
                 num_russian += 1;
             }
         }
-        let mut best_num = 0;
-        if num_dutch > 0 {
-            info!("Found {} dutch resources", num_dutch);
-            if best_num < num_dutch {
-                best_num = num_dutch;
-                best_version = Some(VanillaVersion::DUTCH);
-            }
-        }
-        if num_german > 0 {
-            info!("Found {} german resources", num_german);
-            if best_num < num_german {
-                best_num = num_german;
-                best_version = Some(VanillaVersion::GERMAN);
-            }
-        }
-        if num_italian > 0 {
-            info!("Found {} italian resources", num_italian);
-            if best_num < num_italian {
-                best_num = num_italian;
-                best_version = Some(VanillaVersion::ITALIAN);
-            }
-        }
-        if num_polish > 0 {
-            info!("Found {} polish resources", num_polish);
-            if best_num < num_polish {
-                best_num = num_polish;
-                best_version = Some(VanillaVersion::POLISH);
-            }
+
+        let version = match (num_dutch, num_german, num_italian, num_polish, num_russian) {
+            (n, 0, 0, 0, 0) if n > 0 => Some(VanillaVersion::DUTCH),
+            (0, n, 0, 0, 0) if n > 0 => Some(VanillaVersion::GERMAN),
+            (0, 0, n, 0, 0) if n > 0 => Some(VanillaVersion::ITALIAN),
+            (0, 0, 0, n, 0) if n > 0 => Some(VanillaVersion::POLISH),
+            (0, 0, 0, 0, n) if n > 0 => Some(VanillaVersion::RUSSIAN),
+            _ => None,
+        };
+
+        match version {
+            Some(version) => info!(
+                "Detected exclusively language specific files for language {:?}",
+                version
+            ),
+            None => info!("Detected no language specific files or multiple different languages"),
         }
-        if num_russian > 0 {
-            info!("Found {} russian resources", num_russian);
-            if best_num < num_russian {
-                best_num = num_russian;
-                best_version = Some(VanillaVersion::RUSSIAN);
+
+        Ok(version)
+    }
+
+    /// Guess version by building resource pack and comparing it to other resource
+    /// packs in externalized directory
+    fn guess_by_resource_matching(&self, datadir: &Path) -> GuessResult<Option<VanillaVersion>> {
+        let results = self
+            .get_pack_paths()?
+            .par_iter()
+            .map(|path| self.compare_pack(datadir, path))
+            .collect::<Result<Vec<_>, _>>()?;
```

Comment:
This `collect::<Result<Vec<_>, _>>()?` was confusing me.

I would add a comment saying that it stops early and returns an error (non-deterministic) if any of the parallel calls return an error, or just let everything finish and return a combined error message in case of multiple errors (see [`partition_map`](https://docs.rs/rayon/1.1.0/rayon/iter/trait.ParallelIterator.html#method.partition_map)).

Reply:
I added a comment to that. We don't have a way to return multiple errors from that function yet so `partition_map` seems like it would require quite some overhead (collecting all errors and somehow joining them into a single string).